framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -5073,9 +5074,18 @@ public boolean isValidUse(
         AnnotationMirrorSet tops = qualHierarchy.getTopAnnotations();
         TypeMirror declarationTM = declarationType.getUnderlyingType();
         AnnotationMirrorSet upperBounds = atypeFactory.getTypeDeclarationBounds(declarationTM);
+        AbstractViewpointAdapter vpa =
```

Comment:
The visitor should not adapt types. I think we should change `atypeFactory.getTypeDeclarationBounds` to take what it needs to determine the correct bounds for a particular use.
Then the factory that has a viewpoint adaptor can add the adaptation logic.

Reply:
Okay, let me try to move the adaptation over.